### PR TITLE
Wav2Vec2ProcessorWithLM can return N best hypotheses now

### LINKED ISSUE #22150: Returning n-best hypotheses from Wav2Vec2ProcessorWithLM decoder
### Feature request

Currently, the [Wav2Vec2ProcessorWithLM](https://github.com/huggingface/transformers/blob/ae54e3c3b18bac0832ad62ea9b896dfd52a09850/src/transformers/models/wav2vec2_with_lm/processing_wav2vec2_with_lm.py#L67) decode function returns [only the best hypothesis](https://github.com/huggingface/transformers/blob/ae54e3c3b18bac0832ad62ea9b896dfd52a09850/src/transformers/models/wav2vec2_with_lm/processing_wav2vec2_with_lm.py#L572). Shall we extend its functionality and make it return n-best hypotheses, logit_scores, lm_scores, word_offsets so that people could rescore these hypotheses with a larger LM. 

For example, take a look at [NeMo article](https://docs.nvidia.com/deeplearning/nemo/user-guide/docs/en/main/asr/asr_language_modeling.html#neural-rescoring) regarding the rescoring of n-best hypotheses.

### Motivation

 I suppose many people use n-gram models during the shallow fusion stage, the n-grams models are a good fit during the beam search because they are fast. People perform the rescoring of the n-best hypotheses with a larger LM (using them during the decoding is too slow so it makes sense to apply them during the rescoring of n-best hypotheses that come out of the ASR system). They fuse the score which comes out of the ASR with the perplexity-like score from the LM. If this external model is trained on the domain data it will drastically improve the WER of the resulting model.

### Your contribution

If it sounds like a good feature to you, that can be potentially adopted let me know and I'll prepare the PR 😃

### PR DESCRIPTION
**Pull request by vsokolovskii**
# What does this PR do?

Fixes #22150 , now the user can specify the number of hypotheses which will be returned after the decoding stage. If the specified number is higher than the actual number of hypotheses then all hypotheses will be returned. This is useful when the user wants to run the rescoring on the n-best hypotheses (check out the motivation in the linked issue). Wav2Vec2DecoderWithLMOutput class was already prepared for this feature and [this comment in the code](https://github.com/huggingface/transformers/blob/2355e463955a5392c1acf1964d89747e8b146a6f/src/transformers/models/wav2vec2_with_lm/processing_wav2vec2_with_lm.py#L571) said that this feature will be eventually added, so here it is.

I tried not to break anything that relies on the current version of the decode function, the doc string is updated with a new parameter. All tests passed. The code was well-formatted.

## Before submitting
- [ x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests? 
**Is this necessary for a such small feature?**

@younesbelkada @ArthurZucker  @sanchit-gandhi , does this make sense to you, guys? Is there anything else I should add?

### CODE REVIEW

**Review by ArthurZucker — APPROVED** [review]
LGTM! Thanks for working on this 😉

File: src/transformers/models/wav2vec2_with_lm/processing_wav2vec2_with_lm.py [hunk 1]
```diff
@@ -564,17 +570,37 @@ def decode(
...
+                lm_score=lm_scores[0],
+                word_offsets=word_offsets[0] if word_offsets is not None else None,
+            )
+        else:
+            return Wav2Vec2DecoderWithLMOutput(
+                text=hypotheses[:n_best],
+                logit_score=logit_scores[:n_best],
+                lm_score=lm_scores[:n_best],
+                word_offsets=word_offsets[:n_best] if word_offsets is not None else None,
+            )
```
Comment (amyeroberts) [inline src/transformers/models/wav2vec2_with_lm/processing_wav2vec2_with_lm.py:603]: There's no need to have separate logic here - the slicing will select the 0'th element is `n_best==1` 

```suggestion
        return Wav2Vec2DecoderWithLMOutput(
            text=hypotheses[:n_best],
            logit_score=logit_scores[:n_best],
            lm_score=lm_scores[:n_best],
            word_offsets=word_offsets[:n_best] if word_offsets is not None else None,
        )
```

File: src/transformers/models/wav2vec2_with_lm/processing_wav2vec2_with_lm.py [hunk 2]
```diff
@@ -450,6 +450,7 @@ def decode(
         unk_score_offset: Optional[float] = None,
         lm_score_boundary: Optional[bool] = None,
         output_word_offsets: bool = False,
+        n_best: Optional[int] = 1,
```
Comment (amyeroberts) [inline src/transformers/models/wav2vec2_with_lm/processing_wav2vec2_with_lm.py:453]: As `n_best` cannot be `None` 

```suggestion
        n_best: int = 1,
```

**Review by amyeroberts — APPROVED** [review]
LGTM - thanks for the addition! 

It would be good to have the opinion of @patrickvonplaten before merging to make sure this is in line with the decoding design

File: src/transformers/models/wav2vec2_with_lm/processing_wav2vec2_with_lm.py (hunk 1 above)
Comment (amyeroberts) [reply src/transformers/models/wav2vec2_with_lm/processing_wav2vec2_with_lm.py:603]: That's a good point! Sounds good to me :)

**Comment by sgugger** [comment]
We might have more luck with @sanchit-gandhi ;-)

**Review by patrickvonplaten — COMMENTED** [review]
Thanks a lot for the PR @vsokolovskii,

Just to better understand what happens now in case we decoder a batch of logits with `n_best > 1` - > will we return a list of a list of text in this case? 

Wondering if that's the API that we want - @sanchit-gandhi wdyt?

**Comment by vsokolovskii** [comment]
> Thanks a lot for the PR @vsokolovskii,
> 
> Just to better understand what happens now in case we decoder a batch of logits with `n_best > 1` - > will we return a list of a list of text in this case?
> 
> Wondering if that's the API that we want - @sanchit-gandhi wdyt?

Take a look at the [description of the output class arguments](https://github.com/huggingface/transformers/blob/48327c57182fdade7f7797d1eaad2d166de5c55b/src/transformers/models/wav2vec2_with_lm/processing_wav2vec2_with_lm.py#L45) that you have, you've already prepared everything for this change and I just added the return statement. There should be the possibility to get more than one hypothesis from the ASR in order to rescore it with a larger model, take a look at the motivation section in the linked issue. 🤗

**Review by sanchit-gandhi — APPROVED** [review]
Very cool feature @vsokolovskii! Regarding @patrickvonplaten's question about batch decoding, we don't actually have the argument `n_best` for the `batch_decode` method, it's only for the single-item, `decode` method. So currently, we'd never be returning batches of n-best hypothesis.

WDYT about adding `n_best` to the `batch_decode` method as well @vsokolovskii? In this case, I think we should match the output format to generate's beam search method as `[batches * num_sequences, output_sequences]` (see https://huggingface.co/docs/transformers/internal/generation_utils#transformers.generation.BeamSearchDecoderOnlyOutput.sequences)

**Comment by vsokolovskii** [comment]
@sanchit-gandhi aha... gotcha. Check out the new changes, please.


> Very cool feature @vsokolovskii! Regarding @patrickvonplaten's question about batch decoding, we don't actually have the argument `n_best` for the `batch_decode` method, it's only for the single-item, `decode` method. So currently, we'd never be returning batches of n-best hypothesis.
> 
> WDYT about adding `n_best` to the `batch_decode` method as well @vsokolovskii? In this case, I think we should match the output format to generate's beam search method as `[batches * num_sequences, output_sequences]` (see https://huggingface.co/docs/transformers/internal/generation_utils#transformers.generation.BeamSearchDecoderOnlyOutput.sequences)

**Comment by vsokolovskii** [comment]
@ArthurZucker @amyeroberts , could you please rerun the tests once the pipeline is fixed, I believe that it's not caused by my changes.

**Comment by sgugger** [comment]
The code quality check not passing is not due to your PR at first glance, but to make sure, could you rebase on main? It has been fixed on the main branch.

**Comment by vsokolovskii** [comment]
> The code quality check not passing is not due to your PR at first glance, but to make sure, could you rebase on main? It has been fixed on the main branch.

thanks! forgot yo update my fork
